src/compiler/checker.ts

```diff
@@ -628,31 +650,64 @@ namespace ts {
             Debug.assert(declaration !== undefined, "Block-scoped variable declaration is undefined");
 
             // first check if usage is lexically located after the declaration
-            let isUsedBeforeDeclaration = !isDefinedBefore(declaration, errorLocation);
-            if (!isUsedBeforeDeclaration) {
-                // lexical check succeeded however code still can be illegal.
-                // - block scoped variables cannot be used in its initializers
-                //   let x = x; // illegal but usage is lexically after definition
-                // - in ForIn/ForOf statements variable cannot be contained in expression part
-                //   for (let x in x)
-                //   for (let x of x)
-
-                // climb up to the variable declaration skipping binding patterns
-                let variableDeclaration = <VariableDeclaration>getAncestor(declaration, SyntaxKind.VariableDeclaration);
-                let container = getEnclosingBlockScopeContainer(variableDeclaration);
-
-                if (variableDeclaration.parent.parent.kind === SyntaxKind.VariableStatement ||
-                    variableDeclaration.parent.parent.kind === SyntaxKind.ForStatement) {
-                    // variable statement/for statement case,
-                    // use site should not be inside variable declaration (initializer of declaration or binding element)
-                    isUsedBeforeDeclaration = isSameScopeDescendentOf(errorLocation, variableDeclaration, container);
-                }
-                else if (variableDeclaration.parent.parent.kind === SyntaxKind.ForOfStatement ||
-                    variableDeclaration.parent.parent.kind === SyntaxKind.ForInStatement) {
-                    // ForIn/ForOf case - use site should not be used in expression part
-                    let expression = (<ForInStatement | ForOfStatement>variableDeclaration.parent.parent).expression;
-                    isUsedBeforeDeclaration = isSameScopeDescendentOf(errorLocation, expression, container);
-                }
+            let isUsedBeforeDeclaration = false;
+            switch (getRelativeLocation(declaration, errorLocation)) {
+                case RelativeLocation.DifferentFilesLocatedBefore:
+                    isUsedBeforeDeclaration = true;
+                    break;
+                case RelativeLocation.SameFileLocatedBefore:
+                    // try to detect if forward reference to block scoped variable is inside function
+                    // such forward references are permitted (they are still technically can be incorrect (i.e. in case of IIFEs) 
+                    // but detecting these case is more complicated task)  

```

Comment:
Isn't this true regardless of whether the declaration and reference are in the same file? Seems like it would apply equally to DifferentFilesLocatedBefore